File: trunk/Projects/Framework-Samples/PLINQO/CSharp/Tracker/Source/Tracker.Tests/CacheTests/FromCacheFirstOrDefaultTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using CodeSmith.Data.Caching;
using CodeSmith.Data.Linq;
using NUnit.Framework;
using Tracker.Core.Data;

namespace Tracker.Tests.CacheTests
{
    [TestFixture]
    public class FromCacheFirstOrDefaultTests : RoleTests
    {
        [Test]
        public void SimpleTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var query = db.Role.Where(r => r.Name == "Test Role");
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault();

                var cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                var list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);
            }

        }

        [Test]
        public void DurationTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var query = db.Role.Where(r => r.Name == "Test Role");
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault(2);

                var cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                var list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);

                System.Threading.Thread.Sleep(2500);

                cache = HttpRuntime.Cache.Get(key) as byte[]; 
                Assert.IsNull(cache);
            }

        }

        [Test]
        public void AbsoluteExpirationTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var query = db.Role.Where(r => r.Name == "Test Role");
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault(2);

                var cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                var list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);

                System.Threading.Thread.Sleep(2500);

                cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNull(cache);
            }

        }

        [Test]
        public void SlidingExpirationTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var query = db.Role.Where(r => r.Name == "Test Role");
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault(new CacheSettings(TimeSpan.FromSeconds(2)));

                var cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                var list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);

                System.Threading.Thread.Sleep(1500);

                cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);

                System.Threading.Thread.Sleep(1500);

                cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id);

                System.Threading.Thread.Sleep(2500);

                cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNull(cache);
            }

        }

        [Test]
        public void NoCacheEmptyResultTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var guid = System.Guid.NewGuid().ToString();
                var query = db.Role.Where(r => r.Name == guid);
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault(new CacheSettings(2) { CacheEmptyResult = false });

                Assert.IsNull(role);

                var cache = HttpRuntime.Cache.Get(key);
                Assert.IsNull(cache);
            }

        }

        [Test]
        public void CacheEmptyResultTest()
        {
            using (var db = new TrackerDataContext {Log = Console.Out})
            {
                var guid = System.Guid.NewGuid().ToString();
                var query = db.Role.Where(r => r.Name == guid);
                var key = query.Take(1).GetHashKey();
                var role = query.FromCacheFirstOrDefault(new CacheSettings(2) { CacheEmptyResult = true });

                Assert.IsNull(role);

                var cache = HttpRuntime.Cache.Get(key) as byte[];
                Assert.IsNotNull(cache);

                var list = cache.ToCollection<Role>();
                Assert.IsNotNull(list);
                Assert.AreEqual(0, list.Count);
            }

        }
    }
}

File: Projects/Framework-Samples/PLINQO/VisualBasic/Tracker/Source/Tracker.Tests/CacheTests/FromCacheFirstOrDefaultTests.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading
Imports System.Web
Imports CodeSmith.Data.Caching
Imports CodeSmith.Data.Linq
Imports NUnit.Framework
Imports Tracker.Core.Data
Imports Guid = System.Guid

Namespace Tracker.Tests.CacheTests
    <TestFixture()> _
    Public Class FromCacheFirstOrDefaultTests
        Inherits RoleTests
        <Test()> _
        Public Sub SimpleTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = "Test Role")
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault()

                Dim cache As Byte() = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                Dim list As ICollection(Of Role) = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)
            End Using
        End Sub

        <Test()> _
        Public Sub DurationTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = "Test Role")
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault(2)

                Dim cache As Byte() = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                Dim list As ICollection(Of Role) = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)

                Thread.Sleep(2500)

                cache = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNull(cache)
            End Using
        End Sub

        <Test()> _
        Public Sub AbsoluteExpirationTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = "Test Role")
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault(2)

                Dim cache As Byte() = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                Dim list As ICollection(Of Role) = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)

                Thread.Sleep(2500)

                cache = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNull(cache)
            End Using
        End Sub

        <Test()> _
        Public Sub SlidingExpirationTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = "Test Role")
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault(New CacheSettings(TimeSpan.FromSeconds(2)))

                Dim cache As Byte() = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                Dim list As ICollection(Of Role) = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)

                Thread.Sleep(1500)

                cache = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                list = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)

                Thread.Sleep(1500)

                cache = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                list = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(role.Id, list.FirstOrDefault().Id)

                Thread.Sleep(2500)

                cache = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNull(cache)
            End Using
        End Sub

        <Test()> _
        Public Sub NoCacheEmptyResultTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim guid__1 As String = Guid.NewGuid().ToString()
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = guid__1)
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault(New CacheSettings(2) With { _
                 .CacheEmptyResult = False _
                })

                Assert.IsNull(role)

                Dim cache As Object = HttpRuntime.Cache.[Get](key)
                Assert.IsNull(cache)
            End Using
        End Sub

        <Test()> _
        Public Sub CacheEmptyResultTest()
            Using db As New TrackerDataContext() With { _
             .Log = Console.Out _
            }
                Dim guid__1 As String = Guid.NewGuid().ToString()
                Dim query As IQueryable(Of Role) = db.Role.Where(Function(r) r.Name = guid__1)
                Dim key As String = query.Take(1).GetHashKey()
                Dim role As Role = query.FromCacheFirstOrDefault(New CacheSettings(2) With { _
                 .CacheEmptyResult = True _
                })

                Assert.IsNull(role)

                Dim cache As Byte() = TryCast(HttpRuntime.Cache.[Get](key), Byte())
                Assert.IsNotNull(cache)

                Dim list As ICollection(Of Role) = cache.ToCollection(Of Role)()
                Assert.IsNotNull(list)
                Assert.AreEqual(0, list.Count)
            End Using
        End Sub
    End Class
End Namespace
